File: code/generate_strategies/IR(pseudocode)/cpp/cpp/function_marshmallow_load.cpp

#include <string>
#include <nlohmann/json.hpp>

using ordered_json = nlohmann::ordered_json;

class ProductSchema {
public:
    std::string deserialize_product_data(const std::string &data) {
        try {
            ordered_json parsed = ordered_json::parse(data);
            if (!parsed.contains("name") || !parsed.contains("price")) {
                return "Error: invalid input";
            }
            if (!parsed["name"].is_string() || !parsed["price"].is_number()) {
                return "Error: invalid input";
            }

            ordered_json result;
            result["name"] = parsed["name"];
            result["price"] = parsed["price"];
            if (parsed.contains("in_stock")) {
                if (!parsed["in_stock"].is_boolean()) {
                    return "Error: invalid input";
                }
                result["in_stock"] = parsed["in_stock"];
            } else {
                result["in_stock"] = true;
            }
            return result.dump();
        } catch (ordered_json::exception &) {
            return "Error: invalid input";
        }
    }
};

File: manual/data/java/src/function_marshmallow_load.java
package com.example;

import com.google.gson.*;
import java.util.Locale;

public class ProductSchema {
    public String deserialize_product_data(String data) {
        try {
            // 配置Gson
            JsonParser parser = new JsonParser();
            JsonElement jsonElement = parser.parse(data);

            if (!jsonElement.isJsonObject()) {
                return "Error: invalid input";
            }

            JsonObject jsonObject = jsonElement.getAsJsonObject();

            // 验证必需字段
            if (!jsonObject.has("name") || !jsonObject.has("price")) {
                return "Error: invalid input";
            }

            // 获取name
            JsonElement nameElement = jsonObject.get("name");
            if (!nameElement.isJsonPrimitive() || !nameElement.getAsJsonPrimitive().isString()) {
                return "Error: invalid input";
            }
            String name = nameElement.getAsString();

            // 获取price
            JsonElement priceElement = jsonObject.get("price");
            if (!priceElement.isJsonPrimitive() || !priceElement.getAsJsonPrimitive().isNumber()) {
                return "Error: invalid input";
            }
            double price = priceElement.getAsDouble();

            // 处理in_stock的默认值
            boolean inStock = true;
            if (jsonObject.has("in_stock")) {
                JsonElement inStockElement = jsonObject.get("in_stock");
                if (!inStockElement.isJsonPrimitive() || !inStockElement.getAsJsonPrimitive().isBoolean()) {
                    return "Error: invalid input";
                }
                inStock = inStockElement.getAsBoolean();
            }

            // 构建输出，使用String.format确保数字格式正确
            return String.format(Locale.US,
                    "{\"name\":\"%s\",\"price\":%.2f,\"in_stock\":%s}",
                    name,
                    price,
                    inStock);
        } catch (Exception e) {
            return "Error: invalid input";
        }
    }
}